### ISSUE TITLE
HTTP/1 server

### ISSUE THREAD

**Issue by divchenko** [facebook/fbthrift#99]
I'm trying to get a setup where I have two server endpoints: one to expose new high-performance ThriftServer with the Header transport with binary/compact protocol and another one to expose HTTP server with simple JSON protocol. I see that all servers in cpp/server/example/ got deprecated. Although it seems to be possible to make thrift and fbthrift co-exist it'd be great to use a single project (e.g. to leverage C++ 11 features in fbthrift).
Several questions:
1. Is this separation intentional and for a long term you are planning to maintain ThriftServer with the header transport only? It means that I have to use both thrift and fbthrift (looks like I can reuse generated message classes).
2. Any plans to support HTTP/2 (it seems to have same multiplexing features as the Header transport)?

**Comment by andrewcox** [comment]
Have you tried sending an HTTP message to a header server yet?

We don't typically do this here, so I can't say 100% w/o actually setting something up to try it, but I suspect it might work, because THeader.cpp has backward compatibility to fallback to older types of messages including HTTP.

**Comment by roticv** [comment]
Unfortunately, THeader.cpp only supports HTTP clients (limited support), and support for HTTP server isn't present.

**Comment by andrewcox** [comment]
Here is the code for interpreting an HTTP message from a client, on the server:

https://github.com/facebook/fbthrift/blob/master/thrift/lib/cpp/transport/THeader.cpp#L149

It might not work with the async IO setup of ThriftServer, or it might disabled somewhere when the channel is getting setup (support for each client type can be individually enabled/disabled) but it is there.

**Comment by divchenko** [comment]
Thanks, it looks like that I would need to do parsing of HTTP myself, which is fine. Let me try to read more through the code and try to make it work.

**Comment by andrewcox** [comment]
Well, because thriftserver uses async IO, you need to know when there is a whole message, without a frame size (which you won’t have with HTTP messages) the channel needs to define a way to find message boundaries, and it looks like it does not do so (yet). So actually this looks unlikely to work without changes to the server channels.

**Comment by divchenko** [comment]
I've just discovered quite interesting chunk of code
https://github.com/facebook/fbthrift/blob/master/thrift/lib/cpp2/server/Cpp2Connection.cpp#L304
For HTTP/1 I don't care much about performance, I can even open/close sockets on each request. I mean from that callback I can delegate to a separate thread pool and do blocking operations on that socket from then on. I also don't see any examples which use that http handler.
Anyway, this seems to be an overkill so I'd rather use apache.org thrift for HTTP/1 server.

**Comment by divchenko** [comment]
Latest update: I've managed to make HTTP/1 work with ThriftServer in async mode. There is a way to set HTTP handler and take it from there (see the link for Cpp2Connection above). I had to extend THttpParser to provide a server-side implementation (there is only client-side available) and use TAsyncTransport to read/write HTTP data. Everything seems to work nicely: I've managed to keep HTTP/1 connections open and reuse I/O and worker pools of ThirftServer.
There are some confusing things though: ThriftServer::setGetHandler (it should rather be called HTTPHandler), THttpParser::constructHeader (it should rather be called constructResponse), etc.
Are you planning to provide a formal server-side HTTP support or you are planning to keep things the way they are? I know you don't care much about HTTP, but it'd be nice to have if you ever decide to contribute this back to apache.org. I can submit a patch if you want to.
